### DISCUSSION TITLE
How can I introduce a delay for InlineCompletionItemProvider or make it such that the provider is only triggered with a keyboard shortcut??

### DISCUSSION THREAD

**Discussion by chiralcentre** [microsoft/vscode-discussions#684] [Extension Development QnA] [answered]
Currently I have a VSCode extension with working inline completion functionality. However, I would like to make changes to the extension such that the inline completion function is only triggered upon pressing a keyboard key (eg TAB), or minimally introduce a delay for the inline completion feature such that the feature is not automatically triggered for every word I type. I have researched online and it appears that CancellationTokens may be useful, but I am not quite sure how I can use them in this context, or how to link keyboard shortcuts accordingly. Any help will be greatly appreciated. Thank you!

**Comment by tamuratak** [comment, marked as answer]
Refer to `InlineCompletionContext.InlineCompletionTriggerKind`. If it is set to `Automatic`, make the `provideInlineCompletionItems` function return an empty array. This will ensure that inline completions do not appear during normal editing.


```ts
    export enum InlineCompletionTriggerKind {
        /**
         * Completion was triggered explicitly by a user gesture.
         * Return multiple completion items to enable cycling through them.
         */
        Invoke = 0,


        /**
         * Completion was triggered automatically while editing.
         * It is sufficient to return a single completion item in this case.
         */
        Automatic = 1,
    }
```

- https://github.com/microsoft/vscode/blob/6c8cdf325eb1dc8a0e2ea9205a1d2ca05f69c101/src/vscode-dts/vscode.d.ts#L4773-L4785
- https://code.visualstudio.com/api/references/vscode-api#InlineCompletionItemProvider

**Reply by chiralcentre** [reply]
Thank you for your response! I have added a return of an empty array when triggerKind = 1. At the same time, because I cannot find any documentation, do you happen to know what type of user gestures will set triggerKind as 0? In addition, is there a way to link a keyboard shortcut to the extension such that the extension is activated when the keyboard shortcut is pressed?

**Reply by tamuratak** [reply]
> do you happen to know what type of user gestures will set triggerKind as 0?

No.

> is there a way to link a keyboard shortcut to the extension such that the extension is activated when the keyboard shortcut is pressed?

Find the command associated with the shortcut and set the command to `"activationEvents"` in `package.json`. See

https://code.visualstudio.com/api/references/activation-events#onCommand

In your case, the command might be `editor.action.inlineSuggest.trigger`. Hence, 

```json
	"activationEvents": [
		"onCommand:editor.action.inlineSuggest.trigger"
	],
```

**Reply by chiralcentre** [reply]
Thank you for your response! I have successfully binded a keyboard shortcut to activate my VSCode extension. Your help is greatly appreciated.

**Reply by Glopez-a** [reply]
Hi @chiralcentre ,

I'm developing an extension and trying to bind a keyboard shortcut to activate the suggestion. Did you resolve something similar?

I would appreciate some help.

**Reply by chiralcentre** [reply]
I am not sure if I understand your question correctly, but if I am not wrong you are looking for a way to bind a keyboard shortcut for inline code completions? If this is the case, you can go to Command Palette, and set your own custom keyboard shortcut for the Trigger Inline Suggestion command.

**Comment by rhulha** [comment]
```
"contributes": {
		"commands": [
			{
				"command": "extension.command1",
				"title": "My Inline Completion Demo Settings"
			}
		],
		"keybindings": [
			{
				"key": "ctrl+o",
				"command": "editor.action.inlineSuggest.trigger",
				"mac": "ctrl+o"
			}

		]
	},
```
